source/pool/contracts/Pool.sol

```diff
@@ -0,0 +1,161 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/utils/Pausable.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/cryptography/MerkleProof.sol";
+
+/**
+ * @title Pool: Claim Tokens Based on a Pricing Function
+ */
+contract Pool is Ownable, Pausable {
+  using SafeMath for uint256;
+
+  // Higher the scale, lower the output for a claim
+  uint256 public immutable scale;
+
+  // Max percentage for a claim with infinite score
+  uint256 public immutable max;
+
+  // Mapping of tree root to boolean to enable claims
+  mapping(bytes32 => bool) public roots;
+
+  // Mapping of tree root to account to boolean to mark as claimed
+  mapping(bytes32 => mapping(address => bool)) public claimed;
+
+  /**
+   * @notice Events
+   */
+  event Enable(bytes32 root);
+  event Single(bytes32 root, address account, address token, uint256 amount);
+  event Bulk(bytes32[] roots, address account, address token, uint256 amount);
+
+  /**
+   * @notice Structs
+   */
+  struct Claim {
+    bytes32 root;
+    uint256 score;
+    bytes32[] proof;
+  }
+
+  /**
+   * @notice Constructor
+   * @param _scale uint256
+   * @param _max uint256
+   */
+  constructor(uint256 _scale, uint256 _max) public {
+    scale = _scale;
+    max = _max;
+  }
+
+  /**
+   * @notice Enables claims on the merkle tree of a set of scores
+   * @param root bytes32
+   */
+  function enable(bytes32 root) external onlyOwner {
+    require(roots[root] == false, "ROOT_EXISTS");
+    roots[root] = true;
+    emit Enable(root);
+  }
+
+  /**
+   * @notice Single claim of tokens to be transferred to msg.sender
+   * @param root bytes32
+   * @param score uint256
+   * @param proof bytes32[]
+   * @param token address
+   */
+  function single(
+    bytes32 root,
+    uint256 score,
+    bytes32[] memory proof,
+    address token
+  ) public {
+    require(roots[root], "ROOT_NOT_ENABLED");
+    require(!claimed[root][msg.sender], "CLAIM_INVALID");
+    require(verify(msg.sender, root, score, proof), "PROOF_INVALID");
+    claimed[root][msg.sender] = true;
+
+    uint256 amount = getOutput(score, token);
+    IERC20(token).transfer(msg.sender, amount);
+    emit Single(root, msg.sender, token, amount);
+  }
+
+  /**
+   * @notice Bulk claim of tokens to be transferred to msg.sender
+   * @param claims Claim[]
+   * @param token address
+   */
+  function bulk(Claim[] memory claims, address token) public {
```

Comment:
It feels a little bit inconsistent to use Claim here but using the Claim fields directly in `single`. One option would be to also use the Claim struct in `single`, or to just remove `single` and rename `bulk` to `claim` and use that for everything